packages/@yoda/bluetooth/hfp.js

```diff
@@ -0,0 +1,295 @@
+ */
+class BluetoothHfp extends EventEmitter {
+  constructor (deviceName) {
+    super()
+    this.lastMsg = {
+      'hfpstate': 'invalid', // opened, open failed, closing, closed, invalid.
+      'connect_state': 'invalid', // connecting, connected, connect failed, disconnecting, disconnected, invalid.
+      'connect_address': null,
+      'connect_name': null,
+      'service': 'active', // inactive, active.
+      'call': 'inactive', // inactive, active.
+      'setup': 'none', // none, incoming, outgoing, alerting.
+      'held': 'none', // none, hold_active, hold.
+      'audio': 'off' // on, off.
+    }
+    this._end = false
+    this.localName = deviceName // local bluetooth device name
+
+    this._flora = new FloraComp(logger)
+    this._flora.handlers = {
+      'bluetooth.hfp.event': this._onEvent.bind(this)
+    }
+    this._flora.init(null, {
+      'uri': 'unix:/var/run/flora.sock',
+      'bufsize': 40960,
+      'reconnInterval': 10000
+    })
+  }
+}
+
+/**
+ * @private
+ */
+BluetoothHfp.prototype.matchState = function (msg, filter) {
+  return (filter.hfpstate === msg.hfpstate || filter.hfpstate === undefined) &&
+    (filter.connect_state === msg.connect_state || filter.connect_state === undefined) &&
+    (filter.service === msg.service || filter.service === undefined) &&
+    (filter.call === msg.call || filter.call === undefined) &&
+    (filter.setup === msg.setup || filter.setup === undefined) &&
+    (filter.held === msg.held || filter.held === undefined) &&
+    (filter.audio === msg.audio || filter.audio === undefined)
+}
+
+/**
+ * @private
+ */
+BluetoothHfp.prototype._onEvent = function (data) {
+  if (this._end) {
+    logger.warn(`Already destroied!`)
+    return
+  }
+  var msg = JSON.parse(data[0] + '')
+  logger.debug(`on event(action:${msg.action})`)
+
+  if (msg.action === 'stateupdate') {
+    var last = this.lastMsg
+    logger.debug(`last: ${last.hfpstate} ${last.connect_state} ${last.call} ${last.setup} ${last.held} ${last.audio}`)
+    logger.debug(`now:  ${msg.hfpstate} ${msg.connect_state} ${msg.call} ${msg.setup} ${msg.held} ${msg.audio}`)
+    if (this.matchState(msg, this.lastMsg)) {
+      logger.warn(`Received same msg!`)
+      return
+    }
+    if (msg.setup === 'alerting') {
+      this.lastMsg.setup = 'none'
+      logger.warn(`Ignore alerting msg to avoid disturb incoming & outgoing!`)
+      return
+    }
+    this.lastMsg = Object.assign(this.lastMsg, msg)
+    var stateHit = false
+    stateFilters.forEach((filter) => {
+      if (this.matchState(msg, filter.inflowMsg)) {
+        var event = filter.outflowEvent
+        logger.debug(`Match ${event.type}.${event.state}`)
+        this.emit(event.type, event.state)
+        stateHit = true
+      }
+    })
+    if (!stateHit) {
+      logger.warn(`Mismatch state, please check state-mapping!`)
+    }
+  } else if (msg.action === 'ring') {
+    // We ignore 'ring' and use incoming to indicate ringing state instead.
```

Comment:
Why do we ignore the ring?

Reply:
I have already given the explanation: // We ignore 'ring' and use incoming to indicate ringing state instead.

Reply:
After check, I think framework should send `ring` event to app. Whether ignore it or use it, should determined by application itself.